File: apex_dashboard/apex_dashboard/page/executive_control_center/executive_control_center.py
from __future__ import annotations

from typing import Dict, List, Optional

import frappe
from frappe import _
from frappe.utils import getdate, nowdate

from apex_dashboard.dashboard import utils as dashboard_utils


@frappe.whitelist()
def get_context(context):
	context.no_cache = 1
	context.page_title = _("Executive Control Center")
	return context


@frappe.whitelist()
def get_dashboard_data(company: Optional[str] = None, posting_date: Optional[str] = None):
	try:
		posting_date = posting_date or nowdate()
		posting_date = str(getdate(posting_date))

		data = build_executive_snapshot(company=company, posting_date=posting_date)

		return {
			"success": True,
			"filters": {
				"company": company,
				"posting_date": posting_date,
			},
			"data": data,
		}
	except Exception as exc:
		frappe.log_error(frappe.get_traceback(), "Executive Control Center - get_dashboard_data")
		return {
			"success": False,
			"error": _(str(exc)),
		}


def build_executive_snapshot(company: Optional[str], posting_date: str) -> Dict[str, object]:
	account_map = dashboard_utils.get_dashboard_account_groups().get("executive_control_center", {})
	kpi_map: Dict[str, object] = account_map.get("kpis", {}) if isinstance(account_map, dict) else {}

	total_cash_entry = _prepare_total_cash_entry(kpi_map, company, posting_date)
	net_working_capital_entry = _prepare_working_capital_entry(kpi_map, company, posting_date)
	net_profit_entry = _prepare_net_profit_entry(kpi_map, company, posting_date)
	critical_entry = _prepare_critical_commitments_entry(kpi_map, company, posting_date)

	alerts = _generate_alerts(
		total_cash_entry,
		net_working_capital_entry,
		critical_entry,
	)

	shortcuts = [
		{"label": _("السيولة والتمويل"), "route": ["page", "cash-liquidity-dashboard"]},
		{"label": _("الذمم والتحصيل"), "route": ["page", "receivables-dashboard"]},
		{"label": _("الالتزامات والضرائب"), "route": ["page", "liabilities-dashboard"]},
		{"label": _("الأداء الربحي"), "route": ["page", "pl-performance-dashboard"]},
	]

	return {
		"kpis": [
			total_cash_entry,
			net_working_capital_entry,
			net_profit_entry,
			critical_entry,
		],
		"alerts": alerts,
		"shortcuts": shortcuts,
	}


def _prepare_total_cash_entry(kpi_map, company: Optional[str], posting_date: str) -> Dict[str, object]:
	accounts = []
	if isinstance(kpi_map, dict):
		accounts = kpi_map.get("total_cash") or []

	balances = dashboard_utils.get_balances_for_accounts(accounts, company=company, posting_date=posting_date)
	return {
		"key": "total_cash",
		"label": _("إجمالي السيولة"),
		"totals": {
			"by_currency": dashboard_utils.summarize_balances_by_currency(balances),
			"base": dashboard_utils.summarize_base_total(balances),
		},
		"accounts": accounts,
		"balances": [balance.__dict__ for balance in balances],
		"indicator": "positive",
	}


def _prepare_working_capital_entry(kpi_map, company: Optional[str], posting_date: str) -> Dict[str, object]:
	assets = []
	liabilities = []

	if isinstance(kpi_map, dict):
		wc_map = kpi_map.get("net_working_capital") or {}
		if isinstance(wc_map, dict):
			assets = wc_map.get("assets") or []
			liabilities = wc_map.get("liabilities") or []

	asset_balances = dashboard_utils.get_balances_for_accounts(assets, company=company, posting_date=posting_date)
	liability_balances = dashboard_utils.get_balances_for_accounts(
		liabilities, company=company, posting_date=posting_date
	)

	asset_base = dashboard_utils.summarize_base_total(asset_balances)
	liability_base = dashboard_utils.summarize_base_total(liability_balances)
	net_base = asset_base - liability_base

	return {
		"key": "net_working_capital",
		"label": _("صافي رأس المال العامل"),
		"totals": {
			"assets": dashboard_utils.summarize_balances_by_currency(asset_balances),
			"liabilities": dashboard_utils.summarize_balances_by_currency(liability_balances),
			"base": net_base,
		},
		"accounts": {
			"assets": assets,
			"liabilities": liabilities,
		},
		"balances": {
			"assets": [balance.__dict__ for balance in asset_balances],
			"liabilities": [balance.__dict__ for balance in liability_balances],
		},
		"indicator": "info" if net_base >= 0 else "danger",
	}


def _prepare_net_profit_entry(kpi_map, company: Optional[str], posting_date: str) -> Dict[str, object]:
	accounts = []
	if isinstance(kpi_map, dict):
		accounts = kpi_map.get("net_profit") or []

	balances = dashboard_utils.get_balances_for_accounts(accounts, company=company, posting_date=posting_date)
	total_base = dashboard_utils.summarize_base_total(balances)

	return {
		"key": "net_profit",
		"label": _("صافي الربحية التراكمية"),
		"totals": {
			"by_currency": dashboard_utils.summarize_balances_by_currency(balances),
			"base": total_base,
		},
		"accounts": accounts,
		"balances": [balance.__dict__ for balance in balances],
		"indicator": "positive" if total_base >= 0 else "warning",
	}


def _prepare_critical_commitments_entry(kpi_map, company: Optional[str], posting_date: str) -> Dict[str, object]:
	accounts = []
	if isinstance(kpi_map, dict):
		accounts = kpi_map.get("critical_commitments") or []

	balances = dashboard_utils.get_balances_for_accounts(accounts, company=company, posting_date=posting_date)
	total_base = dashboard_utils.summarize_base_total(balances)

	return {
		"key": "critical_commitments",
		"label": _("الالتزامات الحرجة"),
		"totals": {
			"by_currency": dashboard_utils.summarize_balances_by_currency(balances),
			"base": total_base,
		},
		"accounts": accounts,
		"balances": [balance.__dict__ for balance in balances],
		"indicator": "danger" if total_base > 0 else "info",
	}


def _generate_alerts(
	total_cash_entry: Dict[str, object],
	net_working_capital_entry: Dict[str, object],
	critical_entry: Dict[str, object],
) -> List[Dict[str, object]]:
	alerts: List[Dict[str, object]] = []

	total_cash = (total_cash_entry.get("totals") or {}).get("base", 0)
	net_wc = (net_working_capital_entry.get("totals") or {}).get("base", 0)
	critical = (critical_entry.get("totals") or {}).get("base", 0)

	if total_cash is not None and critical is not None and total_cash < critical:
		alerts.append(
			{
				"level": "danger",
				"message": _("السيولة الحالية أقل من الالتزامات الحرجة، يُرجى مراجعة خطة التمويل."),
			}
		)

	if net_wc is not None and net_wc < 0:
		alerts.append(
			{
				"level": "warning",
				"message": _("صافي رأس المال العامل بالسالب، قد تواجه ضغطًا في التمويل قصير الأجل."),
			}
		)

	if not alerts:
		alerts.append(
			{
				"level": "info",
				"message": _("لا توجد تنبيهات حرجة حالياً."),
			}
		)

	return alerts

File: apex_dashboard/apex_dashboard/page/executive_control_center/executive_control_center.js
frappe.provide("apex_dashboard.dashboard");

frappe.pages["executive-control-center"].on_page_load = function (wrapper) {
	frappe.require(
		[
			"assets/apex_dashboard/js/dashboard_common.js",
			"assets/apex_dashboard/css/dashboard_common.css",
		],
		() => init_executive_page(wrapper)
	);
};

function init_executive_page(wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __("Executive Control Center"),
		single_column: true,
	});

	const state = {
		filters: {
			company: frappe.defaults.get_user_default("company"),
			posting_date: frappe.datetime.get_today(),
		},
		data: null,
	};

	const layout = build_layout(page);
	setup_filter_bar(page, state, layout, refresh);

	function refresh() {
		page.set_indicator(__("جارِ التحميل..."), "blue");
		frappe
			.call("apex_dashboard.apex_dashboard.page.executive_control_center.executive_control_center.get_dashboard_data", {
				company: state.filters.company,
				posting_date: state.filters.posting_date,
			})
			.then((response) => {
				const payload = response.message || {};
				if (!payload.success) {
					throw new Error(payload.error || __("فشل تحميل البيانات"));
				}
				state.data = payload.data || {};
				state.filters = Object.assign({}, state.filters, payload.filters || {});
				render_dashboard(layout, state);
				page.set_indicator(__("جاهز"), "green");
			})
			.catch((error) => {
				console.error(error);
				page.set_indicator(__("خطأ"), "red");
				frappe.msgprint({
					title: __("خطأ"),
					message: error.message || __("تعذر تحميل بيانات لوحة التحكم التنفيذية."),
					indicator: "red",
				});
			});
	}

	refresh();
}

function build_layout(page) {
	const $container = $(page.body);
	$container.empty();

	const $wrapper = $(`
		<div class="gt-dashboard-wrapper">
			<section class="gt-dashboard-section">
				<div class="gt-dashboard-cards" data-role="kpi-cards"></div>
			</section>
			<section class="gt-dashboard-section gt-dashboard-section--split">
				<div class="gt-dashboard-alerts" data-role="alerts"></div>
				<div class="gt-dashboard-shortcuts" data-role="shortcuts"></div>
			</section>
		</div>
	`);

	$container.append($wrapper);

	return {
		kpis: $wrapper.find('[data-role="kpi-cards"]')[0],
		alerts: $wrapper.find('[data-role="alerts"]')[0],
		shortcuts: $wrapper.find('[data-role="shortcuts"]')[0],
	};
}

function setup_filter_bar(page, state, layout, refresh) {
	const company_field = page.add_field({
		fieldtype: "Link",
		fieldname: "company",
		label: __("شركة"),
		options: "Company",
		default: state.filters.company,
		change() {
			state.filters.company = company_field.get_value();
			refresh();
		},
	});

	const date_field = page.add_field({
		fieldtype: "Date",
		fieldname: "posting_date",
		label: __("تاريخ"),
		default: state.filters.posting_date,
		change() {
			state.filters.posting_date = date_field.get_value();
			refresh();
		},
	});

	page.add_action_icon("refresh", () => refresh());
}

function render_dashboard(layout, state) {
	render_kpis(layout.kpis, state.data);
	render_alerts(layout.alerts, state.data);
	render_shortcuts(layout.shortcuts, state.data);
}

function render_kpis(container, data) {
	if (!container) return;
	const kpis = (data && data.kpis) || [];
	const cards = kpis.map((kpi) => {
		const totals = (kpi && kpi.totals) || {};
		const baseValue = totals.base || 0;
		const indicator = kpi.indicator || (baseValue >= 0 ? "positive" : "danger");
		const byCurrency = totals.by_currency || {};
		const subtitle = Object.keys(byCurrency)
			.map((currency) => DashboardCommon.formatCurrency(byCurrency[currency], currency))
			.join(" • ");

		return {
			title: kpi.label || "",
			value: DashboardCommon.formatCurrency(
				baseValue,
				frappe.defaults.get_default("currency") || frappe.sys_defaults.currency || "EGP"
			),
			subtitle,
			indicator,
			onClick: () => open_related_accounts(kpi),
		};
	});

	DashboardCommon.renderCards(container, cards);
}

function render_alerts(container, data) {
	if (!container) return;
	const alerts = (data && data.alerts) || [];

	container.innerHTML = "";

	if (!alerts.length) {
		container.innerHTML = `<div class="gt-dashboard-empty">${__("لا توجد تنبيهات")}</div>`;
		return;
	}

	alerts.forEach((alert) => {
		const item = document.createElement("div");
		item.className = `gt-dashboard-alert level-${alert.level || "info"}`;
		item.innerHTML = `
			<span class="gt-dashboard-alert__badge">${(alert.level || "").toUpperCase()}</span>
			<span class="gt-dashboard-alert__message">${frappe.utils.escape_html(alert.message || "")}</span>
		`;
		container.appendChild(item);
	});
}

function render_shortcuts(container, data) {
	if (!container) return;
	const shortcuts = (data && data.shortcuts) || [];
	container.innerHTML = "";

	if (!shortcuts.length) {
		container.innerHTML = `<div class="gt-dashboard-empty">${__("لا توجد روابط سريعة")}</div>`;
		return;
	}

	shortcuts.forEach((shortcut) => {
		const button = document.createElement("button");
		button.className = "btn btn-default gt-dashboard-shortcut";
		button.innerText = shortcut.label || "";
		button.addEventListener("click", () => {
			if (Array.isArray(shortcut.route)) {
				frappe.set_route(...shortcut.route);
			}
		});
		container.appendChild(button);
	});
}

function open_related_accounts(kpi) {
	if (!kpi) return;
	const dialog = new frappe.ui.Dialog({
		title: kpi.label || __("الحسابات المرتبطة"),
		fields: [
			{
				fieldname: "html",
				fieldtype: "HTML",
			},
		],
		primary_action_label: __("إغلاق"),
		primary_action: () => dialog.hide(),
	});

	const wrapper = document.createElement("div");
	wrapper.className = "gt-dashboard-dialog";
	dialog.fields_dict.html.$wrapper.empty().append(wrapper);

	const balances = kpi.balances || [];

	if (Array.isArray(balances)) {
		const tableContainer = document.createElement("div");
		wrapper.appendChild(tableContainer);
		DashboardCommon.renderGroupTable(tableContainer, balances);
	} else if (balances && typeof balances === "object") {
		Object.entries(balances).forEach(([label, rows]) => {
			const section = document.createElement("div");
			section.className = "gt-dashboard-dialog__section";
			const heading = document.createElement("h5");
			heading.className = "gt-dashboard-dialog__heading";
			let headingText = label;
			if (label === "assets") {
				headingText = __("الأصول");
			} else if (label === "liabilities") {
				headingText = __("الالتزامات");
			}
			heading.innerText = headingText;
			section.appendChild(heading);

			const tableContainer = document.createElement("div");
			section.appendChild(tableContainer);
			const tableRows = Array.isArray(rows) ? rows : [];
			DashboardCommon.renderGroupTable(tableContainer, tableRows);
			wrapper.appendChild(section);
		});
	} else {
		wrapper.innerHTML = `<div class="gt-dashboard-empty">${__("لا توجد بيانات مرتبطة")}</div>`;
	}

	dialog.show();
}

